cmd/internal/cli/actions_linux.go

```diff
@@ -30,9 +34,62 @@ import (
 	"github.com/sylabs/singularity/internal/pkg/sylog"
 	"github.com/sylabs/singularity/internal/pkg/util/env"
 	"github.com/sylabs/singularity/internal/pkg/util/exec"
+	"github.com/sylabs/singularity/internal/pkg/util/fs"
 	"github.com/sylabs/singularity/internal/pkg/util/user"
 )
 
+func convertImage(filename string, unsquashfsPath string) (string, error) {
+	img, err := image.Init(filename, false)
+	if err != nil {
+		return "", fmt.Errorf("could not open image %s: %s", filename, err)
+	}
+	defer img.File.Close()
+
+	// squashfs only
+	if img.Partitions[0].Type != image.SQUASHFS {
+		return "", fmt.Errorf("not a squashfs root filesystem")
+	}
+
+	// create a reader for rootfs partition
+	reader, err := image.NewPartitionReader(img, "", 0)
+	if err != nil {
+		return "", fmt.Errorf("could not extract root filesystem: %s", err)
+	}
+	s := unpacker.NewSquashfs()
+	if !s.HasUnsquashfs() && unsquashfsPath != "" {
+		s.UnsquashfsPath = unsquashfsPath
+	}
+
+	// keep compatibility with v2
+	tmpdir := os.Getenv("SINGULARITY_LOCALCACHEDIR")
+	if tmpdir == "" {
+		pw, err := user.GetPwUID(uint32(os.Getuid()))
+		if err != nil {
+			return "", fmt.Errorf("could not find current user information: %s", err)
+		}
+		tmpdir = filepath.Join(pw.Dir, ".singularity", "tmp")
+		if !fs.IsDir(tmpdir) {
+			if err := os.Mkdir(tmpdir, 0755); err != nil {
+				return "", fmt.Errorf("could not create directory %s: %s", tmpdir, err)
+			}
+		}
+	}
+
+	// create temporary sandbox
+	dir, err := ioutil.TempDir(tmpdir, "rootfs-")
```

Comment:
I'm questioning this approach. It should be possible to have a shared cache of splatted out images (I believe this was the behavior in 2.x, correct me if I'm wrong anyone). Even if this isn't the case with 2.x, I still think it should be possible:

1. Aquire shared lock on a file representing the hash of the fs we want to splat
2. If we're first to aquire this lock (i.e. create this file) then splat the image out to disk
3. After image extraction is complete, write a second file (or write some data to hash file) to signal that splatted image can be used
4. Aquire exclusive lock on file (block other files from opening/shared locking)
5. Check if any other shared locks on this file exist
6. Delete splatted filesystem if no other shared locks exist
7. Release both exclusive and shared lock

The approach outlined above does seem much more complex, so perhaps we should merge this Pull Request as is and discuss the possibility of a shared cache and revisit if applicable

Reply:
The PR approach is strictly identical to 2.x, create a temporary folder and delete it after container execution, those images were not cached in 2.x.

Now if we go this way, imo it should be handled by cache package instead and additionally deal with multi-node cases with the concern that some network filesystem doesn't support file locking